torch/utils/weak.py

```diff
@@ -0,0 +1,79 @@
+import weakref
+from collections.abc import MutableMapping
+from typing import Dict
+
+
+__all__ = ['WeakTensorRefKey', 'WeakTensorKeyDictionary']
+
+
+# Utility classes for working with weak references to tensors
+
+# torch.Tensors cannot be used as a key in a dictionary
+# because they define a custom __eq__ function which when used
+# to resolve hash collisions will throw when comparing tensors:
+# "RuntimeError: bool value of Tensor with more than one value is ambiguous."
+# To avoid that, we use an object which will hold a Tensor and use
+# its id for both hashing and equality.
+# In order to use this as a weak key reference, we cannot
+# simply use weakref.WeakKeyDictionary because the newly constructed
+# WeakTensorRefKey only use would be a dictionary so it would have no strong
+# references.
+# To get around this issue, we can use it as a normal key, and then set
+# `weakref.finalize` to delete the key when its contained tensor dies.
+
+
+class WeakTensorRefKey(object):
+    def __init__(self, ten):
+        self.ten = weakref.ref(ten)
+        # store id since as soon as ten is deallocated
```

Comment:
so, id is a memory address, and it feels a little sketchy to store a memory address like this, especially since:

`ctypes.cast(some_id, ctypes.py_object).value` is valid.

Instead, can we use something else for identification? or at least, make it abundantly clear that this id might be trashed memory?

Reply:
This is also from the old implementation!

Reply:
I don't think you can ever have a WeakTensorRef key with old memory (or an ABA problem). We always clear entries when the tensor in question dies, so this prevents a stale id from being on the map.

Reply:
+1 to what Ed said. can't be trashed memory.